### DISCUSSION TITLE
How is GDAL installed?

### DISCUSSION THREAD

**Discussion by lewfish** [torchgeo/torchgeo#857] [General]
Hello, I'm trying to understand how `torchgeo` handles dependencies and CI. My motivation is that I'm trying to upgrade [Raster Vision](https://github.com/azavea/raster-vision)'s dependencies and Docker image, and want to see how other projects do this, and we use a similar set of dependencies. It all makes sense to me except for one thing. The [tests workflow](https://github.com/microsoft/torchgeo/blob/main/.github/workflows/tests.yaml#L28-L50) uses a vanilla Ubuntu image and then `pip` installs all of `torchgeo`'s dependencies which [includes `rasterio`](https://github.com/microsoft/torchgeo/blob/main/requirements/required.txt#L15). Rasterio has GDAL as a native dependency which needs to be [installed manually on Linux](https://rasterio.readthedocs.io/en/latest/installation.html#linux). (I've verified that the Rasterio documentation here is correct.)  However, it appears that GDAL is not installed within the test workflow, which means that `rasterio` should raise an error when it is imported. But that does not seem to be happening, so what am I missing?

**Comment by calebrob6** [comment]
Hey @lewfish, the recent (I'm not sure when this started) `rasterio` wheels come with gdal built in. E.g. you can download the windows wheel from PyPI and see that it includes a gdal DLL:
![image](https://user-images.githubusercontent.com/1088078/196518330-e7c403de-909a-4da9-82d7-be8969ada612.png)

**Comment by adamjstewart** [comment]
The [link you shared above](https://rasterio.readthedocs.io/en/latest/installation.html#linux) is only for installing from source. As long as you're installing from a pre-built binary (wheel), GDAL is already included in the binary, as @calebrob6 mentioned. I think this is the case since rasterio 0.17, at least for macOS.

Conda and Spack instead install from source, so GDAL is needed as a formal dependency and built before building rasterio.

**Comment by lewfish** [comment]
Thank you. You are right that GDAL is included in the wheels. I was confused because I am testing this on an ARM64 platform, and `rasterio` doesn't have wheels built for that platform. So, it was building it from source, which does require GDAL to be installed first.

**Reply by lewfish** [reply]
Think twice about buying a Macbook M1 :)

**Reply by adamjstewart** [reply]
Yep, I'm also on an M1. I personally use Spack for this which builds GDAL for me, but Conda or Homebrew should also be able to install GDAL for you and then rasterio can be built from source.
